bblocks/import_tools/imf.py

```diff
@@ -0,0 +1,175 @@
+""" """
+from __future__ import annotations
+
+import pandas as pd
+import requests
+from datetime import datetime
+from bs4 import BeautifulSoup
+import os
+from typing import Optional
+from dataclasses import field
+import warnings
+
+from bblocks.config import PATHS
+from bblocks.cleaning_tools.clean import clean_numeric_series
+from bblocks.import_tools.common import ImportData
+
+
+def _parse_sdr_links(url: str, concat_url: str) -> dict:
+    """Function to parse SDR tables from html
+
+    When called it finds the SDR tables and finds a relevant
+    date and link
+
+    Args:
+        url (str): URL to parse
+        concat_url: base url onto which a href is attached
+
+    Returns:
+        A dictionary with dates as keys and URLs as values
+    """
+
+    try:
+        content = requests.get(url).content
+    except ConnectionError:
+        raise ConnectionError(f'Could not read page: {url}')
+
+    soup = BeautifulSoup(content, 'html.parser')
+    table = soup.find_all('table')[4]
+    links = table.find_all('a')
+
+    return {link.string: f'{concat_url}{link.get("href")}' for link in links}
+
+
+def _get_tsv_url(url: str) -> str:
+    """retrieves the URL for the SDR TSV file
+
+    Args:
+        url (str): URL to parse
+
+    Returns:
+        A link to the TSV file
+    """
+
+    try:
+        content = requests.get(url).content
+    except ConnectionError:
+        raise ConnectionError(f'could not read page: {url}')
+    soup = BeautifulSoup(content, "html.parser")
+    href = soup.find_all("a", string="TSV")[0].get("href")
+
+    return f"https://www.imf.org/external/np/fin/tad/{href}"
+
+
+def _read_sdr_tsv(url: str) -> pd.DataFrame:
+    """Reads a TSV file with SDR data from the web to a pandas dataframe
+
+    Args:
+        url (str): link to the TSV file
+
+    Returns:
+        A pandas dataframe with SDR data
+    """
+
+    df = pd.read_csv(url, delimiter="/t", engine="python").loc[3:]
+    df = df["SDR Allocations and Holdings"].str.split("\t", expand=True)
+    df.columns = ["member", "holdings", "allocations"]
+
+    return (df.melt(id_vars='member', value_vars=["holdings", "allocations"])
+            .pipe(clean_numeric_series, series_columns="value")
+            .rename(columns={"variable": "indicator"})
+            .reset_index(drop=True)
+            )
+
+
+def _check_indicators(indicators: str | list) -> list:
+    """ """
+
+    accepted_indicators = ['allocations', 'holdings']
+    if isinstance(indicators, str):
+        if indicators not in accepted_indicators:
+            raise ValueError(f"{indicators} is not a valid indicator")
+        return [indicators]
+    elif isinstance(indicators, list):
+        for indicator in indicators:
+            if indicator not in accepted_indicators:
+                raise ValueError(f"{indicator} is not a valid indicator")
+        return indicators
+    elif indicators is None:
+        return ['allocations', 'holdings']
+
```

Comment:
this could be simplified quite a bit. There's a bit too much nesting, which makes it harder to follow.

A clearer version would return or raise in one go. For example

```python

if indicators is None:
    return accepted_indicators

if isinstance(indicators, str):
    indicators = [indicators]

if not isinstance(indicators, list):
    raise ValueError(f"{indicators} is not in a valid format (str or list)")

for indicator in indicators:
    if indicator not in accepted_indicators:
        raise ValueError(f"{indicator} is not a valid indicator")

return indicators

```